server/app/controllers/admin/AdminImportController.java

```diff
@@ -114,28 +129,172 @@ public Result saveProgram(Http.Request request) {
       return notFound("Program import is not enabled");
     }
 
+    ErrorAnd<ProgramMigrationWrapper, String> deserializeResult = getDeserializeResult(request);
+    if (deserializeResult.isError()) {
+      return ok(
+          adminImportViewPartial
+              .renderError(deserializeResult.getErrors().stream().findFirst().orElseThrow())
+              .render());
+    }
+
+    ProgramMigrationWrapper programMigrationWrapper = deserializeResult.getResult();
+    ImmutableList<QuestionDefinition> questionsOnJson = programMigrationWrapper.getQuestions();
+    ProgramDefinition programOnJson = programMigrationWrapper.getProgram();
+
+    ProgramDefinition updatedProgram = programOnJson;
+
+    if (questionsOnJson != null) {
+      ImmutableMap<Long, QuestionDefinition> questionsOnJsonById =
+          questionsOnJson.stream()
+              .collect(ImmutableMap.toImmutableMap(QuestionDefinition::getId, qd -> qd));
+
+      ImmutableMap<String, QuestionDefinition> updatedQuestionsMap =
+          updateEnumeratorQuestionsAndSave(questionsOnJson, questionsOnJsonById);
+      ImmutableList<BlockDefinition> updatedBlockDefinitions =
+          updateBlockDefinitions(programOnJson, questionsOnJsonById, updatedQuestionsMap);
+      updatedProgram =
+          programOnJson.toBuilder().setBlockDefinitions(updatedBlockDefinitions).build();
+    }
+
+    programRepository.insertProgramSync(
+        new ProgramModel(updatedProgram, versionRepository.getDraftVersionOrCreate()));
+
+    return ok(adminImportView.render(request));
+  }
+
+  private ErrorAnd<ProgramMigrationWrapper, String> getDeserializeResult(Http.Request request) {
     Form<AdminProgramImportForm> form =
         formFactory
             .form(AdminProgramImportForm.class)
             .bindFromRequest(request, AdminProgramImportForm.FIELD_NAMES.toArray(new String[0]));
 
     String jsonString = form.get().getProgramJson();
 
-    ErrorAnd<ProgramMigrationWrapper, String> deserializeResult =
-        programMigrationService.deserialize(jsonString);
+    return programMigrationService.deserialize(jsonString);
+  }
 
-    if (deserializeResult.isError()) {
-      return ok(
-          adminImportViewPartial
-              .renderError(deserializeResult.getErrors().stream().findFirst().orElseThrow())
-              .render());
+  private ImmutableMap<String, QuestionDefinition> updateEnumeratorQuestionsAndSave(
```

Comment:
Question for reviewer(s): Do you think all these helper methods are fine here or should they be in the `ProgramMigrationService` class?

Reply:
I think it's fine either way but I'd slightly lean toward putting them in the service

Reply:
Leaving them here for now since it's just a slight preference and it's a bit of work to move them. But I'll move them in the next issue.